### PR TITLE
Implement extraction of model fitting parameters

### LINKED ISSUE #241: Access fitting plugin model parameter maps from cubeviz helper
This is a follow-on from #187 .  That provides a very nice plugin to fit models across a whole cube.  But what it does *not* provide is a way to easily access the resulting cube maps from the notebook (or other plugins) for further analysis by the user.

Proposed solution: add an accessor in the `Cubeviz` helper class that the #187 plugin writes into.  This could take several forms - just a raw attribute in the helper, a dictionary mapping the data name that the *original* cube was to its derived parameter maps, or some more glue-integrated approach (a la #240 ?) where the accessor is a property that uses that mechanism, etc.

### PR DESCRIPTION
**Pull request by javerbukh**
Fixes #241 and extends the work @ibusko started in #263.

Model objects can now be accessed by the following commands in any config that includes the `ConfigHelper` mixin:
`specviz.fitted_models`
`models = cubeviz.get_models(model_label="Model", x=10)`
`cubeviz.get_model_parameters(models)`

One example notebook and one concept notebook have been updated to document the changes, although more documentation can be added if this approach is approved.

### CODE REVIEW

File: jdaviz/core/helpers.py [hunk 1]
```diff
@@ -35,3 +35,49 @@ def __init__(self, app=None):
...
+
+        Returns
+        -------
+        parameters : dict
+            dict of Quantity arrays, or None.
+        """
+        all_models = {}
+        if hasattr(self.app, '_fitted_1d_models'):
+            all_models = {**all_models, **self.app._fitted_1d_models}
+        if hasattr(self.app, '_fitted_3d_model'):
```
Comment (pllim) [inline jdaviz/core/helpers.py:52]: Not sure if priority is an issue, but second `if` shouldn't run blindly.
```suggestion
        elif hasattr(self.app, '_fitted_3d_model'):
```

File: jdaviz/core/helpers.py [hunk 2]
```diff
@@ -35,3 +35,49 @@ def __init__(self, app=None):
...
+            all_models = {**all_models, **self.app._fitted_3d_model}
+
+        return all_models
+
+    def get_1d_models(self):
```
Comment (pllim) [inline jdaviz/core/helpers.py:57]: Interpreting what @eteq said this morning, I was thinking more like:

```python
def get_models(ndim=1):
    attrname = f'_fitted_{ndim}d_models'
    return getattr(self, attrname, None)
```

**Review by pllim — COMMENTED** [review]
Drive-by review, FWIW.

File: jdaviz/core/helpers.py (hunk 2 above)
Comment (javerbukh) [reply jdaviz/core/helpers.py:57]: Ah, ok, I think that makes more sense!

Comment (pllim) [reply jdaviz/core/helpers.py:57]: p.s. `fitted_models` probably can call `get_models` somehow to avoid code duplication.

Comment (javerbukh) [reply jdaviz/core/helpers.py]: Thank you for the suggestion!

File: jdaviz/core/helpers.py [hunk 3]
```diff
@@ -35,3 +35,41 @@ def __init__(self, app=None):
...
+            dict of Quantity arrays, or None.
+        """
+        MAX_DIMENSIONS = 5
+        all_models = {}
+
+        for dim in range(1, MAX_DIMENSIONS):
+            attrname = f'_fitted_{dim}d_models'
+            model = getattr(self.app, attrname, None)
+            if model:
+                all_models = {**all_models, **model}
```
Comment (pllim) [inline jdaviz/core/helpers.py:56]: I am not still very sure what the intention is here. Do you want to return on first match, get only the last match, or collate all matches?

Comment (javerbukh) [reply jdaviz/core/helpers.py:56]: Collate all matches.

Comment (javerbukh) [reply jdaviz/core/helpers.py:56]: Maybe this functionality does not need to exist. What do you think @eteq ?

Comment (pllim) [reply jdaviz/core/helpers.py:56]: I just don't see how it is collating with `=` assignment at every iteration. Do you mean to update the dictionary instead?

Comment (pllim) [reply jdaviz/core/helpers.py:56]: Oh wait... you are passing in itself on the right... Nvm... I think maybe using `.update()` method is cleaner, not sure.

Comment (javerbukh) [reply jdaviz/core/helpers.py]: You're right, much easier to understand.

**Review by eteq — CHANGES_REQUESTED** [review]
This is subtley different from what I was thinking: this is returning the model *parameters*, but the goal here is to return the model *objects* (which have the models embedded inside them). So I was imagining `fitted_models` should return a dictionary that maps keys to model *objects*, and then the getter function just iterates over those and returns those with a given dimensionality.

*However*, now that I look at this I see that's rather different then what the current 3d attribute gives. That array is currently confusingly named because it is called `3d_models` instead of `2d_array_of_model_parameters` - that is, the "3rd" dimension is over model parameters, whereas the other dimensions are the data cube dimensions.

So here's my rethink of what we want here: we still want `fitted_models`, and it should be expanded to include the models that speviz creates. But the getter should be instead something like `get_model_patameter_list`, which should create *on the fly* the parameter array, building it from an array of model objects. For specciz models that'll be kinda silly  because you can always just do `helper.fitted_models['my_model'].parameters`, but for larger dimensional datasets the method is highly desirable.

I'm not sure if all of the above should go in this PR vs a first step just getting `fitted_models` working and a second follow-on implementing the helper method, though. What do you think @javerbukh ?

**Comment by pllim** [comment]
Model parameter itself is an object, right? You have to be careful -- If user get them and then modify them, will that in turn change the underlying model object tied to them? I don't know enough of the `modeling` internals to be sure.

**Comment by javerbukh** [comment]
@eteq Let me know if I have this correct: leave `fitted_models` the same (I'm not sure what you mean by " expanded to include the models that specviz creates") and change `get_models` to `get_model_patameter_list` that returns a dict of models `.parameters` rather than just models. Is that right?

@pllim You make a good point and I also do not know exactly how this works. @rosteen can you shed some light on this?

**Comment by rosteen** [comment]
I had a whole long reply written up about this, and then realized that the crux of the issue that I was forgetting about is that the fitting backend for the parallelized cube fitting returns a dictionary of model parameter arrays and the fitted cube, rather than the fitted astropy model object itself and the fitted spectrum in the 1D case. I assume this was done either because saving the actual model object for each spaxel would take too much memory, or because the resulting astropy model object weren't serializable in a way compatible with the parallelization. I'm going to do a quick check to see if it would actually work to return an array of models instead, which would perhaps greatly simplify things.

**Comment by rosteen** [comment]
I submitted a pull request above to this branch that starts implementing the stuff below. Based on a discussion @eteq and I had offline I think the updated ask for this is:

1) All fitted models go into `app.fitted_models` keyed on model label (implemented in my PR to your fork). The helper `fitted_models` can pretty much just be a wrapper around `app.fitted_models`.

2) There's a `get_model_parameters` method that contains the logic to rebuild the parameter arrays that Ivo's code in the fitting backend used to create/return for 3D models (I left that code in but commented out for now in my PR for reference). For 1D models the parameters dict just contains single values instead of arrays. I imagine that this could either return the parameters for all models, or take a model label as input and return only the parameters (or parameter arrays for 3D) for that model. 

The upshot of storing all the models in `app.fitted_models` is that those should contain all the information needed for (2). Probably the most annoying part will be figuring out the most sensible way to index all the individual spaxel models in the 3D case - maybe what I did in my PR is good enough.

Did I miss anything, @eteq ?

**Comment by javerbukh** [comment]
@rosteen I'm not sure what you mean by `get_model_parameters`. Is this a method you created for testing or does this already exist in the app somewhere?

So if I understand correctly, after wrapping `app.fitted_models` into the helper, the next step is be able to separate those models by their dimensionality and return models of `ndim` dimension in the getter, right?

**Comment by rosteen** [comment]
Sorry, I worded that poorly. I meant 1 and 2 are the two "asks" as in those are the things you should code/should exist when you're done. I did the first half of 1 in my commits, we still need the helper wrappers of `app.fitted_models` and a helper `get_model_parameters` method. I don't think we actually need anything that returns only a certain dimensionality of model, being able to index on model label is enough.

**Comment by eteq** [comment]
> Did I miss anything, @eteq ?

Nope, I think that covers it for this PR! Just one clarification on 1: the `fitted_models` parameter should have keys be the label and values be model *objects* (e.g. `GaussianModel1D` or similar - i.e., subclasses of `astropy.modeling.Model`). 

We talked a bit about the idea that *later* we may want to implement some other accessors beyond `get_model_parameters` that might do more complicated stuff like building numpy dtype=object arrays that match `get_model_parameters`, but the idea would be this would use information already embedded in `fitted_models` and the model objects, rather than storing additional information outside of the model objects themselves like what's currently in there.

**Comment by javerbukh** [comment]
The latest commit implements the `get_models()` and `get_model_parameters()` methods. The SpecvizExample notebook and concepts/Cube_fit notebook can be used for testing this PR.
Pinging for review @eteq @rosteen @ibusko

File: jdaviz/core/helpers.py [hunk 4]
```diff
@@ -35,3 +39,197 @@ def __init__(self, app=None):
...
+            else:
+                find_label = model_label
+
+            # If x and y is set, return keys that match the model plus that
+            # coordinate pair. If only x or y is set, return keys that fit
+            # that value for the appropriate coordinate. If neither is set,
+            # add the model object to the selected_models dict with the label
+            # as the key.
+            if x and y:
+                find_label = f"{find_label} ({x}, {y})"
```
Comment (rosteen) [inline jdaviz/core/helpers.py:102]: Pretty sure these parens need escapes as well for the regex. Currently supplying both x and y values returns an empty dict for me.
```suggestion
                find_label = f"{find_label} \({x}, {y}\)"
```

File: jdaviz/core/helpers.py [hunk 5]
```diff
@@ -35,3 +39,197 @@ def __init__(self, app=None):
...
+            elif x:
+                find_label = r"{} \({}, .+\)".format(find_label, x)
+            elif y:
+                find_label = f"{find_label} (.+, {y})"
```
Comment (rosteen) [inline jdaviz/core/helpers.py:106]: Here too, although specifying y was working for me. Maybe better to be consistent anyway.
```suggestion
                find_label = f"{find_label} \(.+, {y}\)"
```

File: jdaviz/core/helpers.py [hunk 6]
```diff
@@ -35,3 +39,197 @@ def __init__(self, app=None):
...
+            else:
+                selected_models[label] = models[label]
```
Comment (rosteen) [inline jdaviz/core/helpers.py:108]: This is making it so all models are returned if no x or y are specified, even if a model label is specified. I think this case should be deleted altogether.

**Review by rosteen — CHANGES_REQUESTED** [review]
Made some suggestions that I think would fix the bugs I'm seeing with `get_models`. From my initial look at `get_model_parameters` I think it looks good, although I think it should be possible to input a model label to `get_model_parameters` as well in order to get back the parameters for just that model (rather than having to go through `get_models` first and then provide the dict of models to `get_model_parameters`). I might do a bit more poking at `get_model_parameters` to make sure I can't break it.

**Review by pllim — COMMENTED** [review]
I only did a general review and this LGTM at a glance. I don't think I am qualified to do a detailed review, given I am not familiar with the internals.

File: jdaviz/core/helpers.py (hunk 6 above)
Comment (javerbukh) [reply jdaviz/core/helpers.py]: Good call!

File: jdaviz/core/helpers.py [hunk 7]
```diff
@@ -35,3 +39,197 @@ def __init__(self, app=None):
...
+                find_label = label
+            else:
+                find_label = model_label
+
+            # If x and y is set, return keys that match the model plus that
+            # coordinate pair. If only x or y is set, return keys that fit
+            # that value for the appropriate coordinate. If neither is set,
+            # add the model object to the selected_models dict with the label
+            # as the key.
+            if x and y:
```
[the remark this answers is not in the document]
Comment (javerbukh) [reply jdaviz/core/helpers.py]: Good catch!

**Comment by javerbukh** [comment]
@rosteen I implemented your suggestions for `get_models()`. I'm still not sure about having a `model_label` parameter in `get_model_parameters()` since that will require duplicating code from `get_models()`. I'll work on getting that in and see if it will be a simple addition, but that has not been included in the latest commit.

**Comment by rosteen** [comment]
Rather than duplicating code, couldn't you just call `get_models` from inside `get_model_parameters` in that case, passing through the `model_label` input to get the right dictionary?

**Comment by javerbukh** [comment]
Good call...I guess my brain is still warming up today 😅

File: jdaviz/core/helpers.py [hunk 8]
```diff
@@ -35,3 +39,197 @@ def __init__(self, app=None):
...
+        Returns
+        -------
+        :dict: a dictionary of the form
+            {model name: {parameter name: [[`astropy.units.Quantity`]]}}
+            for 3d models or
+            {model name: {parameter name: `astropy.units.Quantity`}} where the
+            Quantity object represents the parameter value and unit of one of
+            spaxel models or the 1d models, respectively.
+        """
+        if not models and model_label:
```
Comment (rosteen) [inline jdaviz/core/helpers.py:134]: It's probably worth covering the case where both `model` and `model_label` were input, since it's possible.

Comment (javerbukh) [reply jdaviz/core/helpers.py]: Good call, done.

**Comment by rosteen** [comment]
I think this is almost there, the parameters retrieval looks good. There's one more case in the model retrieval that I'm trying to think through - currently if you have  two models labeled e.g. "Model" and a "Model 2" and do get_model(model_labels="Model"), it will return both models, whereas I would expect it to return only the exact match. Could you refactor slightly to do an exact match on the portion of the model label before the (x, y) part? I'm imagining something like

```
if model_label is not None:
    if label.split(" (")[0] != model_label:
        continue
```

right at the start before the other cases. Note that the split still works if " (" isn't in the label, it just returns the whole string unsplit in [0].

**Comment by javerbukh** [comment]
@rosteen Done!

**Review by rosteen — APPROVED** [review]
Every case I can think of is now working as I would expect, thanks for all the work on this!

**Comment by rosteen** [comment]
You may want to edit the first comment to show the updated usage for the next reviewer.

**Review by duytnguyendtn — APPROVED** [review]
Ok, thanks to the work done this week, I now have a solid understanding of what's going on to be confident in giving this review. Looks like I can retrieve the models and the associated parameters appropriately.

I did have a comment about the potential redundancy of "fitted_models" vs "get_models" but I remembered reading somewhere of the reason why. I'm not sure if it's just brain fog, but I can't seem to find that comment. Can @rosteen or @javerbukh link me to that message for posterity's sake? Thanks :)

**Comment by rosteen** [comment]
`fitted_models` is the app-level data structure where the models are stored, `get_models` is a helper-level convenience method to retrieve subsets of those models.

**Comment by rosteen** [comment]
That made me realize that the docstring for the helper parameter that hooks into the app `fitted_models` is out of date. I'll create a followup issue.

**Comment by javerbukh** [comment]
Good call @rosteen ! As for your question @duytnguyendtn , the intention was that `fitted_models` allows some ability to directly access certain models with `cubeviz.fitted_models["Model (1, 2)"]`, although that same functionality can be duplicated with `cubeviz.get_models(model_label="Model", x=1, y=2)`. Maybe there are possible use cases with being able to change/replace the models using `fitted_models`? I'm not sure, maybe it can be removed completely if there's really no use for it.

**Comment by duytnguyendtn** [comment]
Nope that makes sense! It didn't register in my head that `get_models` was a convenience method because usually it abstracts a `viz.app.some_function()` call, whereas in this case, both `fitted_models` and `get_models` are at the helper level. But considering @rosteen's comment, it makes sense for there to be a convenience method, since `fitted_models` is just the data structure container itself. I probably I would've picked it up if it were `viz.app.fitted_models` 😅

**Pull request merged**